### ISSUE TITLE
How to use adaptive learning rate to train models?

### ISSUE THREAD

**Issue by wangg12** [dmlc/mxnet#540]
I found `lr_scheduler`, but I could't find any scheduler. I wonder what kind of scheduler I can use in mxnet? For example, if I need to train a model by decaying the learning rate of SGD every few epoches, how could I make it?

**Comment by jermainewang** [comment]
Hi,

Currently, we only have normal [SGD optimizer](https://github.com/dmlc/mxnet/blob/master/python/mxnet/optimizer.py#L86). One way to have decaying learning rate is to implement a new Optimizer like adagrad or adam. Would you like to have a try?

Another way to walkaround this is to use the `callback` interface in `model.fit`. Please see [`callback.py`](https://github.com/dmlc/mxnet/blob/master/python/mxnet/callback.py) for some examples. You could implement an `epoch_end_callback` and pass it to `model.fit` to adjust learning rate after several epochs. Please let us know what is your choice. Thank you very much.

Best,
Minjie

**Comment by antinucleon** [comment]
Mingjie, in misc I implemented lr schedule

**Comment by jermainewang** [comment]
I see, like @antinucleon said, you could directly used code in [misc.py](https://github.com/dmlc/mxnet/blob/master/python/mxnet/misc.py).

**Comment by wangg12** [comment]
Well, I see. And I want to know that whether mxnet will provide adadelta, adam, adagrad or other optimizing methods and more flexible learning rate schedulers?

**Comment by shuokay** [comment]
Cant call ```mx.misc.FactorScheduler```
For example, in lenet.py,:
```
model.fit(X=train, eval_data=val,
          batch_end_callback=mx.callback.Speedometer(100),
          epoch_end_callback=[mx.callback.do_checkpoint(model_prefix), mx.misc.FactorScheduler(2,0.9)]
)
```
and python complain
```
/usr/local/lib/python2.7/dist-packages/mxnet-0.5.0-py2.7.egg/mxnet/model.pyc in _train_multi_device(symbol, ctx, arg_names, param_names, aux_names, arg_params, aux_params, begin_epoch, end_epoch, epoch_size, optimizer, kvstore, update_on_kvstore, train_data, eval_data, eval_metric, epoch_end_callback, batch_end_callback, logger, work_load_list)
    383             if isinstance(epoch_end_callback, list):
    384                 for call in epoch_end_callback:
--> 385                     call(epoch, symbol, arg_params, aux_params)
    386             else:
    387                 epoch_end_callback(epoch, symbol, arg_params, aux_params)

TypeError: __call__() takes exactly 2 arguments (5 given)

```

**Comment by zachmayer** [comment]
@jermainewang How do you use a callback to adjust the model parameters?  Could you give a short example `epoch_end_callback` that adjusts the learning rate?

**Comment by jermainewang** [comment]
Hi,

Here is a simple example on how to do it in callback way. I use the [cifar10](https://github.com/dmlc/mxnet/tree/master/example/cifar10) example as the code base.

What you need to do is to capture the optimizer in a function closure:
```python
import mxnet.optimizer as opt
def test_cifar():
  logging.basicConfig(level=logging.DEBUG)
  gpus = [mx.gpu(i) for i in range(num_gpus)]
  sgd_opt = opt.SGD(learning_rate=0.005, momentum=0.9, wd=0.0001, rescale_grad=(1.0/batch_size))
  model = mx.model.FeedForward(ctx=gpus, symbol=softmax, num_epoch=num_epoch,
              optimizer=sgd_opt, initializer=mx.init.Uniform(0.07))
  def lr_callback(param):
    if param.nbatch % 10 == 0:
      sgd_opt.lr /= 10 # decrease learning rate by a factor of 10 every 10 batches
    print 'nbatch:%d, learning rate:%f' % (param.nbatch, sgd_opt.lr)

  model.fit(X=train_dataiter, eval_data=test_dataiter, batch_end_callback=lr_callback)
```
On my machine, it outputs something like followings:
```
[12:58:18] src/io/iter_image_recordio.cc:212: ImageRecordIOParser: data/cifar/train.rec, use 1 threads for decoding..
[12:58:18] src/io/./iter_normalize.h:98: Load mean image from data/cifar/cifar_mean.bin
[12:58:18] src/io/iter_image_recordio.cc:212: ImageRecordIOParser: data/cifar/test.rec, use 1 threads for decoding..
[12:58:18] src/io/./iter_normalize.h:98: Load mean image from data/cifar/cifar_mean.bin
INFO:root:Start training with [gpu(0)]
nbatch:1, learning rate:0.050000
nbatch:2, learning rate:0.050000
nbatch:3, learning rate:0.050000
nbatch:4, learning rate:0.050000
nbatch:5, learning rate:0.050000
nbatch:6, learning rate:0.050000
nbatch:7, learning rate:0.050000
nbatch:8, learning rate:0.050000
nbatch:9, learning rate:0.050000
nbatch:10, learning rate:0.005000
nbatch:11, learning rate:0.005000
nbatch:12, learning rate:0.005000
nbatch:13, learning rate:0.005000
```

Hope this helps,

-Minjie

**Comment by jermainewang** [comment]
@shuokay, the correct way to use `lr_scheduler` is to pass it to `mx.model.FeedForward`:
In lenet.py
```
model = mx.model.FeedForward(ctx=dev, symbol=lenet, num_epoch=20,  
  learning_rate=0.05, momentum=0.9, wd=0.00001, 
  lr_scheduler=mx.misc.FactorScheduler(step=5))
```
Interestingly, there is a bug in optimizer that is fixed in #622, so it should work now.

Thank you for your question!

Minjie

**Comment by udibr** [comment]
you can try my Adam pull request https://github.com/dmlc/mxnet/pull/623
